src/publish.rs

```diff
@@ -1,185 +1,254 @@
 
-        assert_eq!(order.sell_amount, 2);
+        let spread = Spread::from_f64(0.0);
+
+        let order = new_dai_bitcoin_order(wallet, book, btc(2.0), rate, spread);
+
+        assert_eq!(order.sell_amount, btc(2.0));
     }
 
     #[test]
     fn given_an_available_balance_and_fees_sell_balance_minus_fees() {
-        let wallet = Wallet::new(10, 1);
+        let wallet = Wallet::new(btc(10.0), btc(1.0));
+
+        let book = Book::new(btc(2.0));
 
-        let book = Book::new(2);
+        let rate = BigRational::from_f64(1.0).unwrap();
 
-        let order = new_order(wallet, book, 2, 1.0, Spread::new(0.0).unwrap());
+        let spread = Spread::from_f64(0.0);
 
-        assert_eq!(order.sell_amount, 1);
+        let order = new_dai_bitcoin_order(wallet, book, btc(2.0), rate, spread);
+
+        assert_eq!(order.sell_amount, btc(1.0));
     }
 
     #[test]
     fn given_a_rate_return_order_with_both_amounts() {
-        let wallet = Wallet::new(1051, 1);
+        let wallet = Wallet::new(btc(1051.0), btc(1.0));
+
+        let book = Book::new(btc(50.0));
+
+        let rate = BigRational::from_f64(0.1).unwrap();
 
-        let book = Book::new(50);
+        let spread = Spread::from_f64(0.0);
 
-        let order = new_order(wallet, book, 9999, 10.0, Spread::new(0.0).unwrap());
-        // 1 Buy => 10 Sell
-        // ? Buy => 1000 sell
-        // 100 Buy => 1000 Sell
+        let order = new_dai_bitcoin_order(wallet, book, btc(9999.0), rate, spread);
 
-        assert_eq!(order.sell_amount, 1000);
-        assert_eq!(order.buy_amount, 100)
+        // 1 Sell => 0.1 Buy
+        // 1000 Sell => 100 Buy
+        assert_eq!(order.sell_amount, btc(1000.0));
+        assert_eq!(order.buy_amount, dai(100.0));
+
+        let rate = BigRational::from_f64(10.0).unwrap();
+
+        let spread = Spread::from_f64(0.0);
+
+        let order = new_dai_bitcoin_order(wallet, book, btc(9999.0), rate, spread);
+
+        assert_eq!(order.sell_amount, btc(1000.0));
+        assert_eq!(order.buy_amount, dai(10_000.0));
     }
 
     #[test]
     fn given_a_rate_and_spread_return_order_with_both_amounts() {
-        let wallet = Wallet::new(1051, 1);
+        let wallet = Wallet::new(btc(1051.0), btc(1.0));
+
+        let book = Book::new(btc(50.0));
+
+        let rate = BigRational::from_f64(0.1).unwrap();
 
-        let book = Book::new(50);
+        let spread = Spread::from_f64(0.03);
 
-        let order = new_order(wallet, book, 9999, 10.0, Spread::new(0.03).unwrap());
-        // 1 Buy => 10 Sell
-        // ? Buy => 1000 sell
-        // 100 Buy => 1000 Sell
-        // 3% spread
-        // 103 Buy => 1000 Sell
+        let order = new_dai_bitcoin_order(wallet, book, btc(9999.0), rate, spread);
 
-        assert_eq!(order.sell_amount, 1000);
-        assert_eq!(order.buy_amount, 104); // Rounding up taking in account precision loss
+        assert_eq!(order.sell_amount, btc(1000.0));
+        assert_eq!(order.buy_amount, dai(103.0));
```

Comment:
I'm probably not getting it, but from your test output and this assertion, the value you provide is less precise than the value returned by your function. How did you calculate the expected value?